server.py

```diff
@@ -214,13 +217,107 @@ def _parse_languages(languages: str) -> Dict[str, int]:
     return ret
 
 
+def _parse_experience(experience: str) -> List[int]:
+    """
+    Return a string containing integers separated by commas as a list of ints.
+    """
+
+    ret = list()
+    if not experience:
+        return ret
+    for experience_level in experience.split(","):
+        ret.append(int(experience_level))
+    return ret
+
+
+def _get_coach_json(coach_args: JSON) -> CoachJSON:
+    """
+    Return a CoachJSON object corresponding the args from the get coaches
+    request.
+    """
+
+    # Turn all interger arguments into ints.
+    for int_arg in ["birth_year"]:
+        if int_arg in coach_args:
+            coach_args[int_arg] = int(coach_args[int_arg])
+
+    # Turn all list of interger arguments into lists of ints
+    for experience_args in ["need", "rights", "housing"]:
+        if experience_args in coach_args:
+            coach_args[experience_args] = _parse_experience(
+                coach_args[experience_args]
+            )
+
+    # Special case languages as it's a dict.
+    if "languages" in coach_args:
+        coach_args["languages"] = _parse_languages(coach_args["languages"])
+    
+    return coach_args
+
+
+def _include_coach(coach: CoachJSON, coach_filter: CoachJSON) -> bool:
+    """
+    Check whether to include a coach from the db in a repsoonse by filtering on
+    if the coach matches the fields in coach_filter.
+    """
+
+    if not coach_filter:
+        return True
+
+    # Keys where we need coach[key] == coach_filter[key]
+    direct_comparison_keys = [
+        "available",
+        "bio",
+        "birth_year",
+        "gender",
+        "id",
+        "name",
+    ]
+    for key in direct_comparison_keys:
+        if key in coach_filter:
+            if not coach.get(key) == coach_filter[key]:
+                return False
+
+    # Keys where we need coach[key] to be a superset of coach_filter[key]
+    #
+    # Note: this includes cases where the value is a list and is a dict.
+    subset_comparison_keys = [
+        "housing",
+        "languages",
+        "need",
+        "rights",
+    ]
+    for key in subset_comparison_keys:
+        if key in coach_filter:
+            # If value is list, turn into sets for comparison.
+            if type(coach_filter[key]) is list:
+                if not set(coach_filter[key]) <= set(coach.get(key)):
+                    return False
+
+            # If value is dict, use all to check each key:value pair in
+            # coach_filter[key] is in coach[key].
+            elif type(coach_filter[key]) is dict:
+                if not all(
+                    item in coach.get(key).items()
+                    for item in coach_filter[key].items()
+                ):
+                    return False
```

Comment:
I'd suggest either breaking this up to handling lists and dicts separately, or better treat all keys together without hardcoding the field names, e.g.
```python
for key, filter_val in [k, coach_filter[k] for k in coach.fields() if k in coach_filter]:
    if isinstance(k, list):
        ...
    elif isinstance(k, dict):
        ...
    else:
        # Direct comparison, as above.
```

In either case take inspiration from the simplification above :)